Adjust sds types

poiuj opened the pull request:
sds type should be determined based on the size of the underlying buffer, not the logical length of the sds. Currently we truncate the alloc field in case buffer is larger than we can handle. It leads to a mismatch between alloc field and the actual size of the buffer. Even considering that alloc doesn't include header size and the null terminator.

It also leads to a waste of memory with jemalloc. For example, let's consider creation of sds of length 253. According to the length, the appropriate type is SDS_TYPE_8. But we allocate `253 + sizeof(struct sdshdr8) + 1` bytes, which sums to 257 bytes. In this case jemalloc allocates buffer from the next size bucket. With current configuration on Linux it's 320 bytes. So we end up with 320 bytes buffer, while we can't address more than 255.

The same happens with other types and length close enough to the appropriate powers of 2.

The downside of the adjustment is that with allocators that do not allocate larger than requested chunks (like GNU allocator), we switch to a larger type "too early". It leads to small waste of memory. Specifically: sds of length 31 takes 35 bytes instead of 33 (2 bytes wasted) sds of length 255 takes 261 bytes instead of 259 (2 bytes wasted) sds of length 65,535 takes 65,545 bytes instead of 65,541 (4 bytes wasted) sds of length 4,294,967,295 takes 4,294,967,313 bytes instead of 4,294,967,305 (8 bytes wasted)

poiuj commented:
This patch is subset of changes in https://github.com/valkey-io/valkey/pull/453. I think we can merge this independently. https://github.com/valkey-io/valkey/pull/453 depends on this patch and on https://github.com/valkey-io/valkey/pull/476.

madolson commented on src/sds.c:
```diff
@@ -124,8 +110,6 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     s = (char*)sh+hdrlen;
     fp = ((unsigned char*)s)-1;
     usable = usable-hdrlen-1;
-    if (usable > sdsTypeMaxSize(type))
```
The defensive programmer in me wants to replace this with an assert that usable <= the max size.

poiuj replied:
Done

madolson commented on src/sds.c:
```diff
@@ -59,35 +59,21 @@ static inline int sdsHdrSize(char type) {
+    if (sizeof(struct sdshdr5) + string_size + 1 <= 1<<5)
         return SDS_TYPE_5;
-    if (string_size < 1<<8)
+    if (sizeof(struct sdshdr8) + string_size + 1 <= 1<<8)
```
If possible, I'd like to document this better. The `<=` behavior was surprising to convince myself it was correct, because the actual value of ->alloc doesn't include the overhead, so we have 2-9 bytes of space which happens to always be enough room.

poiuj replied:
I agree. Let me try to improve the documentation here.

madolson reviewed (commented):
I think this makes sense to me. The slightly more memory usage on some systems is likely OK since I think we still prefer to use jemalloc in most cases.

ranshid commented:
@poiuj overall looks much better now. I would suggest we include some unit tests to cover different types edge cases (for example the case of old style alloc overflow). I am comfortable with the asserts, but we will probably want to trigger ci and daily before we add this.

After we finish this PR lets go and fix #453 accordingly

ranshid commented:
@poiuj the tests looks fine. 
Regarding the top comment:

>The downside of the adjustment is that with allocators that do not allocate larger than requested chunks (like GNU allocator), we switch to a larger type "too early". It leads to small waste of memory. Specifically: sds of length 31 takes 35 bytes instead of 33 (2 bytes wasted) sds of length 255 takes 261 bytes instead of 259 (2 bytes wasted) sds of length 65,535 takes 65,545 bytes instead of 65,541 (4 bytes wasted) sds of length 4,294,967,295 takes 4,294,967,313 bytes instead of 4,294,967,305 (8 bytes wasted)

Is it really a waste? we already reduced the overflow in the prev implementation and AFAIK we never should have written pass the usable anyhow right?

poiuj commented:
@ranshid your example is correct. It's a 59 bytes gain if we use jemalloc. My top commented is about libc allocator that wouldn't allocate 320 bytes in this case.

ranshid commented:
> @ranshid your example is correct. It's a 59 bytes gain if we use jemalloc. My top commented is about libc allocator that wouldn't allocate 320 bytes in this case.

Oh I get it now. yes. indeed this is not too bad IMO. Please also consider avoid the malloc_size use in realloc

poiuj commented:
@ranshid sure, I plan to get rid of call to `zmalloc_size` in resize as part of #453.

madolson commented on src/sds.c:
```diff
@@ -58,36 +58,42 @@ static inline int sdsHdrSize(char type) {
+ * up to 2^5, 2^8, 2^16, and 2^32 bytes. We assume jemalloc uses these sizes for
+ * allocation. Otherwise jemalloc allocates larger buffer, which length can't be
+ * stored in alloc field. */
+static inline size_t sdsTypeMaxSize(char type) {
```
Smallest of comments, can we move this function back to where it was so we keep the git history.

poiuj replied:
We now use it in `sdsReqType`. Do you prefer to add a forward declaration of `sdsTypeMaxSize` to keep the git history?

madolson replied:
yes

zhulipeng commented on src/sds.c:
```diff
@@ -124,8 +130,7 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     usable = usable-hdrlen-1;
-    if (usable > sdsTypeMaxSize(type))
-        usable = sdsTypeMaxSize(type);
+    assert(usable <= sdsTypeMaxSize(type));
```
You can't do such assertion here, usable is got from allocator and allocator didn't have such kind of guarantee that it will have maximum size. For example, if I requested 236 bytes, the allocator size could be 248 bytes (good) but maybe 264 bytes. I think the guarantee is that the allocated size >= requested size.

poiuj replied:
You're right, the allocator doesn't guarantee a maximum size. But it doesn't give random blocks either. Jemalloc allocates buffers of pre-defined sizes. And SDS types are aligned accordingly.
If this assertion is false, `usable` is incorrect.

poiuj replied:
I thought about jemalloc. I was sure that libc allocator would return a buffer of the exact size. Apparently I was wrong, thank you for noting it. I'll fix it.

madolson reviewed (approved):
LGTM, but would like @lipzhu and @ranshid to signoff as well. This is sensitive code and would like extra eyes on it.

zhulipeng commented on src/sds.c:
```diff
@@ -124,8 +132,14 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     usable = usable-hdrlen-1;
+
+#ifdef USE_JEMALLOC
+    assert(usable <= sdsTypeMaxSize(type));
```
The assertion here is based on the behavior we expected/monitor from jemalloc, it depends on the internal of jemalloc, is there any guarantee from jemalloc? Maybe a potential risk.

ranshid replied:
why can't we always keep the defending part?  (usable = min(usable, sdsTypeMaxSize(type))
AND also maybe we need to consider keep the assert for the usable size correctness, eg:
assert(usable >= initlen);

madolson replied:
```
assert(usable <= sdsTypeMaxSize(type));
assert(initlen <= usable);
```
Are our two invariants. Should we be doing those asserts regardless of the allocator being used?

ranshid replied:
Yes that is a good point:
we want to make sure that in case JEMALLOC is used we will SURELY have **real** usable == usable, in order to satisfy #453, so I guess we do have to keep assert(usable <= sdsTypeMaxSize(type)); for JEMALLOC (for example in order to assert when JEMALLOC was compiled with different bin sizes?).
I think (as @madolson suggested) the defensive option is better at this point.

ranshid replied:
I would still change the code so that:

> #ifndef USE_JEMALLOC
> usable = MIN(usable, sdsTypeMaxSize(type));
> #endif
> assert(usable <= sdsTypeMaxSize(type) && initlen <= usable);

poiuj replied:
I think there is a better solution. We can adjust types after the allocation/reallocation. Previously I thought there are some edge cases that wouldn't work. But after reconsidering the idea, and noting again that `alloc` doesn't include header size and null terminator, I think it's a better way.
With this solution we don't need to take assumptions for correctness. Instead of truncating `usable`, we're going to adjust the header type.
I'll update the PR with this idea.

poiuj commented:
Updated the branch. Includes the new approach with adjusting types. Also rebased on the latest unstable. Unfortunately, missed some style changes during conflict resolution. But it's small stuff that I can fix later before the actual merge.

ranshid commented on src/sds.c:
```diff
@@ -54,23 +54,27 @@ static inline int sdsHdrSize(char type) {
-    if (type == SDS_TYPE_8) return (1 << 8) - 1;
-    if (type == SDS_TYPE_16) return (1 << 16) - 1;
+    if (type == SDS_TYPE_5)
+        return (1<<5) - 1;
```
I think it is better to leave the current format in order to avoid non-logical changes.

lipzhu replied:
`sdsTypeMaxSize` is used to do a protection before set `alloc` metadata for sds,  the `alloc` meaning here equals actual allocator size - header size - 1. Maybe we should also minus the header size for `sdsTypeMaxSize`?

ranshid reviewed (commented):
maybe we can make a function to capture the repeated code?:
```
 /* Adjust type if usable won't fit into alloc. */
    if (type != SDS_TYPE_5 && usable > sdsTypeMaxSize(type)) {
        type = sdsReqType(usable);
        hdrlen = sdsHdrSize(type);
        usable = bufsize - hdrlen - 1;
        assert(usable <= sdsTypeMaxSize(type));
    }
```

ranshid commented:
LGTM but would like to hear @lipzhu

zhulipeng commented on src/unit/test_sds.c:
```diff
@@ -262,5 +263,97 @@ int test_sds(int argc, char **argv, int flags) {
+    x = sdsnewlen(NULL, 252);
+    TEST_ASSERT_MESSAGE("len 252 type", (x[-1] & SDS_TYPE_MASK) >= SDS_TYPE_8);
+#ifdef USE_JEMALLOC
+    TEST_ASSERT_MESSAGE("len 252 sdsAllocSize", sdsAllocSize(x) == s_malloc_size(sdsAllocPtr(x)));
```
I found the warn message from jemalloc manual. 

> Any discrepancy between the requested allocation size and the size reported by malloc_usable_size() should not be depended on, since such behavior is entirely implementation-dependent.

poiuj replied:
We don't depend on discrepancy between the requested allocation size and the size reported by `malloc_usable_size`. Not in the test, not in the code. The test checks that `sdsAllocSize` returns the same value as `s_malloc_size`. It's always true as `sdsAllocSize` is based on the value we got from `malloc_usable_size` before.
In any case, this test checks that our assumption is correct. As we vendor jemalloc, the only case when it can change (theoretically) is when we update jemalloc.

zhulipeng replied:
> It's always true as sdsAllocSize is based on the value we got from malloc_usable_size before.

You are right, so we can open assertion not only for jemalloc?

poiuj replied:
Oh, you're right! This `ifdef` isn't needed anymore. Updated the PR.

lipzhu reviewed (commented):
LGTM, just one concern.
